Indexing system for Table

mdmueller opened the pull request:
This is a first pass at a new system allowing the `Table` class to hold indices, both for improved grouping performance and for searching by column values. Indices are stored in columns rather than the full table, and indices in turn store a reference to the column(s) they act on, so that composite indices work correctly. Some parts of this PR are final and some are still unfinished, so any comments/suggestions would be great.

### Complete
* The `Index` class, which provides a wrapper around the sorting engine
* Modifications to `Column`, which ensure that indices are modified as appropriate when a `Column` is modified
* Modifications to `Table`

### Unfinished
* ~~The `Table.where()` method, particularly finishing up the searching syntax~~
* ~~The actual sorting engine. I have bintrees (which provides fast binary search tree and red-black tree implementations) as an optional dependency, which is replaced by a pure-Python binary search tree class if bintrees isn't present. I also began work on a pure-Python red-black tree and a sorted array class, both unfinished.~~
Experimental parts of this PR have been removed.

@mdboom @taldcroft

Cadair commented:
This is very exciting! I will try and take a look.

taldcroft commented on astropy/table/column.py:
```diff
@@ -271,11 +273,31 @@ def __reduce__(self):
         if isinstance(item, INTEGER_TYPES):
             return self.data[item]  # Return as plain ndarray or ma.MaskedArray
+
+        elif isinstance(item, slice):
```
So there is a little problem here... In #3930 (and #3929 which it superceded) we decided to remove `__getitem__` from the `BaseColumn` class because it slows single item access (and hence iterating over column values) by more than a factor of 10.  This is a fundamentally due to having to call into the Python layer instead of directly to the C version from ndarray.

I don't see any obvious way to get around this.  What do you think @mdboom ?  If not, we may need to just lose indices whenever you do indexing (slice, fancy indexing, whatever) on a `Column`.  We can still do this at the Table level, which is a more likely case where you really want the indices to follow along.

mdboom replied:
Yeah, I think let's just preserve them at the table level and see whether anyone notices that the aren't preserved when slicing columns...

taldcroft commented on astropy/table/index.py:
```diff
@@ -0,0 +1,308 @@
+        else:
+            num_rows = len(columns[0])
+            table = Table(columns + [[x for x in range(num_rows)]])
+            lines = table.group_by([x.name for x in columns])
```
All column attributes should be accessed via the `info` attribute (e.g. `x.info.name`).  This is the mixin-safe idiom.

taldcroft commented on astropy/table/index.py:
```diff
@@ -0,0 +1,308 @@
+            raise ValueError("Cannot create index without at least one column")
+        else:
+            num_rows = len(columns[0])
+            table = Table(columns + [[x for x in range(num_rows)]])
```
Use `np.arange(num_rows)` ?

taldcroft commented on astropy/table/array.py:
```diff
@@ -0,0 +1,184 @@
+import numpy as np
```
Call this module `sorted_array.py` to be more specific and clear.

taldcroft commented on astropy/table/array.py:
```diff
@@ -0,0 +1,184 @@
+    a numpy structured array.
+    '''
+
+    def __init__(self, lines, col_dtypes=None, data=None):
```
The names are a bit confusing or non-informative to me.  I think that `lines` should be more like `key_cols_table` (if I'm understanding).  I really don't get what `data` is, because `data is None` case `data` ends up being a list of columns + the index column, but in the block below you set `self.length = len(data)` => length of the list => n_key_columns + 1.

mdmueller replied:
Ah, thanks for the catch--this made more sense previously when `data` was a structured array, but I forgot about it when I turned `data` into a list of columns. It really should be `self.length = len(data[0])` (the number of elements in each column), and by the way `data` is only used for slicing. I'll change `lines` to `key_cols_table` as well since that seems like a better name.

taldcroft commented:
This looks like a problem:
```
In [40]: t
Out[40]: 
<Table length=3>
  a      b     c  
int32 float32 str1
----- ------- ----
    3     3.0    e
    2     2.0    d
    1     1.0    c

In [44]: t.insert_row(2, (10, 10.0, 'z'))

In [45]: t
Out[45]: 
<Table length=4>
  a      b     c  
int32 float32 str1
----- ------- ----
    3     3.0    e
    2     2.0    d
   10    10.0    z
    1     1.0    c

In [46]: t.indices
Out[46]: [[[ 1  2  3 10], [2 1 0 2]]]
```

mdboom commented on astropy/table/index.py:
```diff
@@ -0,0 +1,308 @@
+                return index
+    return None
+
+class static_indices:
```
Given the name, I would have thought that this prevents the updating of indices as the values change (and then recalculates the index when the `with` block is exited).  Is that something that would be hard to implement?  The use case being "I need to add 1000 rows to a table, and don't want the index recalculated with each addition".  

Even if we don't implement what I just described, maybe we should change the name of this to avoid confusion?

mdmueller replied:
Actually, this context just prevents indices from being copied upon creation of a new `Table` (e.g. `Table(t)`, `t[1:3]`, etc.) You're right that the name probably isn't appropriate, and actually this is somewhat of a hack to improve the performance of `group_by` (which takes many slices). I'm not sure if there are other use cases, but if we keep it, maybe another solution would be to include a parameter signaling whether to avoid updating indices completely or just to avoid copying indices? Something like:
```
with index_mode('modify'):
   # add rows to a table without index changes
with index_mode('copy'):
   # take slices/boolean masks of a table without index copying
```

mdboom replied:
I like what you suggest -- it will allow us to have other indexing contexts in the future...

mdboom commented on astropy/table/table.py:
```diff
@@ -372,6 +381,139 @@ def filled(self, fill_value=None):
+        <left-paren> := "[" | "("
+        <right-paren> := "]" | ")"
+
+        Example: table.where('[col0] in [{0}, {1}) and [col1] = {2}', 1, 5, 3)
```
I'd like to see some doctests here showing examples of usage.  It's a little unclear from the BNF grammar what everything does.

Also, this, like all docstrings, should be in numpydoc format.

mdmueller commented:
Does anyone have any thoughts about the best way to support indices for mixin columns? The issue is that indices are stored and managed in the `Column` class, so mixin indices will require a workaround--maybe we should store mixin indices directly in `Table`? I can't think of a clean way to update indices based on mixin column modifications, though, unless we change mixin classes directly.

taldcroft commented:
The first thought that comes to mind for mixins is to put the indices into the `info` container. That is the new generic easy to handle attributes in a mixin safe way. What do you think?

mdmueller commented:
Makes sense to me, although the real issue is that indices have to be updated every time a column is modified. For example, `t['col'][row] = val` should update all indices in `t['col']`. This functionality is currently built into `Column` but the fact that mixins can be essentially arbitrary makes it more difficult.

taldcroft commented:
@mhvk - this table indexing PR is touching on the whole mixin concept and you might have some ideas.  See the comments above about indexes and mixins.  One idea is to use some kind of mixin class (here "mixin" in the traditional sense) like `SupportsTableIndexing` to indicate that a mixin like `Quantity` has the machinery in `__setitem__` to handle in-place updates.  Another idea is a class attribute like `supports_table_indexing = True`.  The latter is perhaps simpler.  I haven't tested to see which is faster.

@mdmueller - for the canonical case of `Time` objects (for Time series) we are helped by the fact that `Time` objects are immutable and you cannot set them.  For at least this case there is no problem.  But of course we will want to index `Quantity`s.

mhvk commented:
Was just going to comment we could have the mixin on the `info` object by an attribute, or, perhaps easier, by a subclass that provides the required index -- but I see you edited your comment to say the same thing!

taldcroft commented:
What about naming the `index_modes` as `freeze` (instead of `modify`) and `discard_on_copy` instead of `copy`.  The current names seem to be exactly opposite of what I would expect.

mdmueller commented:
Sounds good -- I'll update the names.

taldcroft commented:
Can a context manager temporarily override `__getitem__` for an instance?  I'm thinking:
```
with col.index_mode('copy_on_getitem'):
    col_new = col[i0:i1]
```
where `__getitem__` is your current `getitem`?

mdmueller commented:
Ah, that does seem like a nice idea for convenience. It should actually be pretty simple to override, e.g.
```
In [1]: class A:
   ...:     def __getitem__(self, item):
   ...:         return item
   ...:         

In [2]: a = A()

In [5]: class context:
    def __init__(self, a):
        self.a = a
    def __enter__(self):
        self.func_copy = self.a.__getitem__
        self.a.__getitem__ = lambda x: x*2
    def __exit__(self, exc_type, exc_value, traceback):
        self.a.__getitem__ = self.func_copy
   ...:         

In [7]: a[5]
5

In [8]: with context(a):
   ...:     print(a[5])
   ...:     
10

In [9]: a[5]
5

```

taldcroft commented on astropy/time/core.py:
```diff
@@ -992,6 +996,7 @@ def argsort(self, axis=-1):
         """
         jd_approx = self.jd
         jd_remainder = (self - self.__class__(jd_approx, format='jd')).jd
+        print(jd_approx, jd_remainder)
```
Leftover debug.

mdmueller replied:
Oops, thanks for the catch!

mdboom commented on astropy/table/column.py:
```diff
@@ -1091,28 +1115,40 @@ def insert(self, obj, values, mask=None):
+        out = super(MaskedColumn, self).__getitem__(item)
+        return self.copy_attribs(out)
+
+    def get_item(self, item):
```
This needs a docstring -- and probably a comment about why both it and `__getitem__` exist.

mdboom commented on astropy/table/index.py:
```diff
@@ -0,0 +1,543 @@
+
+Notes
+-----
+    When a Table is initialized from another Table, indices are
```
I think the indentation here is messing up the formatting.

mdboom commented on astropy/table/table.py:
```diff
@@ -205,13 +208,16 @@ def as_array(self):
         self.meta = meta
         self.formatter = self.TableFormatter()
+        self._copy_indices = True # copy indices from this Table by default
+        self._modify_indices = True # modify indices upon Table modification
```
I don't see where this value is getting used.  Just not implemented yet?

mdmueller replied:
This was what I was previously using to create an indexing mode in which indices ignore modifications, but `_modify_indices` can now be deleted -- thanks for the catch.

taldcroft commented on astropy/table/index.py:
```diff
@@ -0,0 +1,630 @@
+    columns : list or None
+        List of columns on which to create an index. If None,
+        create an empty index for purposes of deep copying.
+    impl : type or None
```
I think `engine` might be a more descriptive name.

taldcroft commented on astropy/table/index.py:
```diff
@@ -0,0 +1,630 @@
+    def __new__(cls, *args, **kwargs):
+        self = super(Index, cls).__new__(cls)
+        self.__init__(*args, **kwargs)
+        return SlicedIndex(self, slice(0, 0, None), original=True)
```
Shouldn't it be possible to move all the `SlicedIndex` methods into `Index`, and have `Index.__init__` take `index_slice` (with default=`slice(None)`) and `original` as kwargs?

mdmueller replied:
Yes, as long as `Index` methods are made to transform rows between coordinates correctly. I suppose that might be a better structure.

taldcroft replied:
I don't quite understand why you are initializing with `slice(0, 0, None)` which is a null (empty) slice.

mdmueller replied:
Ah, that's because for "original" slices, the `stop` property is determined by actual column length rather than the input slice (which determines `_stop`). Now that I think about it, it might make more sense to keep `stop` as an ordinary variable and modify it whenever column length changes.

embray replied:
More to the point, I think `SlicedIndex` should be a subclass of `Index`--this `__new__` should check an `index_slice` argument and return a `SlicedIndex` for a non-trivial slice, and otherwise return a plain `Index` (this also avoids having to call `__init__` manually in the `__new__`).

Now, I haven't read all the code closely yet so I don't know if there's a reason `SlicedIndex` needs to wrap an `Index` rather than work as a subclass.  If so that's fine too, but I'd find it confusing that all Indexes are SlicedIndexes.

taldcroft commented on astropy/table/index.py:
```diff
@@ -0,0 +1,630 @@
+from copy import deepcopy
```
Use `from __future__ import division` to be sure of uniform Py2 / Py3 behavior.

taldcroft commented:
All the new files in the table package should include standard header:
```
# Licensed under a 3-clause BSD style license - see LICENSE.rst
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)
```

taldcroft commented on astropy/table/index.py:
```diff
@@ -0,0 +1,630 @@
+        '''
+        if self.original:
+            return rows
+        elif self.step > 0:
```
It should be possible to do this with numpy array operations, e.g.:
```
row0 = row - self.start
if self.step != 1:
    ok = np.mod(row0, self.step) == 0
    row0 = row0[ok]
ok = (self.start <= row0) & (row0 < self.stop)
return row0[ok] // self.step
```
In general you don't want any pure-Python loops or list comprehensions where you could be looping over anything of order the table length (which could be 10^9 or more).

mdmueller replied:
Sounds good, I'll just modify this to work for negative `self.step` as well.

taldcroft commented on astropy/table/sorted_array.py:
```diff
@@ -0,0 +1,286 @@
+
+    Parameters
+    ----------
+    lines : `Table`
```
It's confusing to me to call this `lines` because that implies something like a list of lines (rows).  Maybe `index_table` or even just `table`?

mdmueller replied:
Sure, `index_table` sounds good.

taldcroft replied:
Is there a good reason you need to bundle the row index column with the values column(s)?  It seems like this creates a lot of special handling (e.g. `len(lines.colnames) - 1` or `lines.columns[:-1]`) that ends up making the code harder to read and maintain.  Can it just be split into a `data` attribute and a `row_index` attribute?  

Is there a reason that `data` cannot or should not be a `Table` instead of a list of columns?

mdmueller replied:
Yeah, splitting it that way is probably best. It's necessary to have the row column bundled with the value columns in a Table before performing `lexsort`, but after that we can split it up again into `data` and `row_index` when passing to the engine `__init__`.

Currently `data` is a list of ndarrays rather than Column objects. I could definitely turn it into a Table if that makes the code easier to read/maintain, although my initial thought was that an ndarray will be faster, e.g. since Column has a custom `__setitem__`.

taldcroft replied:
> Yeah, splitting it that way is probably best. It's necessary to have the row column bundled with the value columns in a Table before performing lexsort, but after that we can split it up again into data and row_index when passing to the engine __init__.

Is that because you are worried about making the sort be stable?  Otherwise I don't see why you need the row as part of the sort.

taldcroft replied:
> Currently data is a list of ndarrays rather than Column objects. I could definitely turn it into a Table if that makes the code easier to read/maintain, although my initial thought was that an ndarray will be faster, e.g. since Column has a custom __setitem__.

This strategy doesn't play well with mixin columns because you end up with an ndarray of objects, which will blow up for long tables:
```
t = simple_table(10)
t['t'] = Time(t['b'], format='cxcsec')
type(t['t'])
t.add_index('t')
t.indices[0]
[[<Time object: scale='tt' format='cxcsec' value=1.0>
 <Time object: scale='tt' format='cxcsec' value=2.0>
 <Time object: scale='tt' format='cxcsec' value=3.0>
 <Time object: scale='tt' format='cxcsec' value=4.0>
 <Time object: scale='tt' format='cxcsec' value=5.0>
 <Time object: scale='tt' format='cxcsec' value=6.0>
 <Time object: scale='tt' format='cxcsec' value=7.0>
 <Time object: scale='tt' format='cxcsec' value=8.0>
 <Time object: scale='tt' format='cxcsec' value=9.0>
 <Time object: scale='tt' format='cxcsec' value=10.0>], [0 1 2 3 4 5 6 7 8 9]]
```

mdmueller replied:
It's because the end result has to include the rows in their original positions. For example to create an index on the table
```
<Table length=3>
 col0 col1
int64 str1
----- ----
    3    a
    1    b
    2    c
```
we add a column corresponding to row numbers:
```
<Table length=3>
 col0 col1  col2
int64 str1 int64
----- ---- -----
    3    a     0
    1    b     1
    2    c     2
```
and sort so that the last column is the new index `sorted_data`:
```
In [16]: t[np.lexsort([t['col1'], t['col0']])]
Out[16]: 
<Table length=3>
 col0 col1  col2
int64 str1 int64
----- ---- -----
    1    b     1
    2    c     2
    3    a     0
```

mdmueller replied:
Ah, I hadn't thought about mixins...in that case using a `Table` seems to make more sense.

taldcroft commented on astropy/table/sorted_array.py:
```diff
@@ -0,0 +1,286 @@
+            self.length = len(data[0])
+            self._data = data
+            return
+        elif col_dtypes is not None:
```
In the function docs the input `lines` is listed as not optional.  If so, then why would you ever need to specify `col_dtypes` explicitly?  That can always be found in the input table of index columns.

mdmueller replied:
`col_dtypes` is present only in the case that lines is `None`, which takes place during an index deepcopy (https://github.com/astropy/astropy/pull/3915/files#diff-bb938335053a3190ec096125099093deR342). I should document that.

taldcroft commented on astropy/table/sorted_array.py:
```diff
@@ -0,0 +1,286 @@
+            self.num_cols = len(lines.colnames) - 1
+            col_dtypes = [x.info.dtype for x in lines.columns[:-1].values()]
+
+        self.length = 0 if lines is None else len(lines)
```
But here I see there is a case for `lines is None`.  What does that mean to be creating an index without any corresponding table columns?

mdmueller replied:
See `deepcopy`--the columns are added after initialization.

taldcroft commented on astropy/table/sorted_array.py:
```diff
@@ -0,0 +1,286 @@
+            col_dtypes = [x.info.dtype for x in lines.columns[:-1].values()]
+
+        self.length = 0 if lines is None else len(lines)
+        self._data = [np.zeros(self.length, dtype=d) for d in col_dtypes]
```
If you split out `row_index` from `data` then I think all this just becomes:
```
self._data = data.copy()
self._row_index = row_index
```
(Or something like that).

taldcroft replied:
And it's not completely clear that you even need to copy since the `data` that get passed in are a sorted version of the original columns so it is already a clean copy that isn't going to be manipulated elsewhere.  In general I wouldn't guess that people should be creating `Index` objects themselves so we can afford to be less conservative about copying.

taldcroft commented:
We need to discuss methods that sit between the low-level index object and `where`.  Basically things like pandas' `loc` and `iloc`.

taldcroft commented:
Can we make the table `indices` attribute support get_item by the name or names of the corresponding columns?  I.e.
```
t.add_index('a')
t.add_index(['a', 'b'])
t.indices['a', 'b']  # returns indices[1]
```
I guess you just need to subclass `list` to make this happen.

taldcroft commented:
I think the `.loc` accessor should handle the same inputs as the pandas version:
  http://pandas.pydata.org/pandas-docs/stable/indexing.html#different-choices-for-indexing
The single label case is the most important, but a list or ndarray of labels is useful as well.

taldcroft commented:
There is a pretty bad performance hit when a table is indexed and you do fancy indexing.  Maybe a pure Python loop is getting invoked somewhere?
```
In [78]: t = simple_table(1000000)

In [79]: timeit t[[1,2]]
1000 loops, best of 3: 451 µs per loop

In [80]: timeit t[1:3]
1000 loops, best of 3: 433 µs per loop

In [81]: t.add_index('a')

In [82]: timeit t[[1,2]]
1 loops, best of 3: 282 ms per loop  # <<<====

In [83]: timeit t[1:3]
1000 loops, best of 3: 449 µs per loop
```
There is also about a base 50% performance hit compared to 1.0.4:
```
In [9]: t = simple_table(1000000)

In [10]: timeit t[1:3]
1000 loops, best of 3: 300 µs per loop

In [11]: timeit t[[1,2]]
1000 loops, best of 3: 310 µs per loop
```

mdmueller commented:
That sounds good. I notice that Pandas also allows for setting values via `.loc` and `.iloc`, which I assume will be useful for our case as well, e.g.
```
t.loc[100:] = 100
```

mdmueller commented:
Hmm....the performance hit for ``t[[1, 2]]`` isn't too surprising (though it needs to be mitigated), but I find it very strange that basic slicing is slower than in 1.0.4. I'll look into it.

taldcroft commented:
This result is not what I would expect.
```
In [84]: t = simple_table(3)

In [85]: t['d'] = Time([1,3,2], format='cxcsec')

In [86]: t.add_index('d')

In [94]: t.loc[:]
Out[94]: 
<Table length=0>
  a      b     c     d   
int32 float32 str1 object
----- ------- ---- ------

In [95]: t.loc[Time(1.5, format='cxcsec'):Time(2.5, format='cxcsec')]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-95-08297e5b2da0> in <module>()
----> 1 t.loc[Time(1.5, format='cxcsec'):Time(2.5, format='cxcsec')]

/Users/aldcroft/git/astropy/astropy/table/index.pyc in __getitem__(self, item)
    722 
    723         index = self.indices[key]
--> 724         rows = index.range((item.start,), (item.stop,))
    725         return self.table[rows]
    726 

/Users/aldcroft/git/astropy/astropy/table/index.pyc in range(self, lower, upper)
    488 
    489     def range(self, lower, upper):
--> 490         return self.sliced_coords(self.index.same_prefix_range(lower, upper))
    491 
    492     def same_prefix(self, key):

/Users/aldcroft/git/astropy/astropy/table/index.pyc in same_prefix_range(self, lower, upper, bounds)
    288         lower = lower + tuple((ncols - n) * [a])
    289         upper = upper + tuple((ncols - n) * [b])
--> 290         return self.data.range(lower, upper, bounds)
    291 
    292     def replace(self, row, col_name, val):

/Users/aldcroft/git/astropy/astropy/table/sorted_array.pyc in range(self, lower, upper, bounds)
    130             and the second argument y to an inclusive upper bound.
    131         '''
--> 132         lower_pos = self.find_pos(lower, 0)
    133         upper_pos = self.find_pos(upper, 0)
    134         if lower_pos == len(self.row_index):

/Users/aldcroft/git/astropy/astropy/table/sorted_array.pyc in find_pos(self, key, data, exact)
     63             key_slice = self.cols[i][begin:end] if i < self.num_cols \
     64                         else self.row_index[begin:end]
---> 65             t = np.searchsorted(key_slice, key[i])
     66             # t is the smallest index >= key[i]
     67             if exact and (t == len(key_slice) or key_slice[t] != key[i]):

/Users/aldcroft/anaconda/envs/astropy/lib/python2.7/site-packages/numpy/core/fromnumeric.pyc in searchsorted(a, v, side, sorter)
   1017     except AttributeError:
   1018         return _wrapit(a, 'searchsorted', v, side, sorter)
-> 1019     return searchsorted(v, side, sorter)
   1020 
   1021 

/Users/aldcroft/git/astropy/astropy/time/core.pyc in __len__(self)
   1284         if self.isscalar:
   1285             raise TypeError("Scalar {0} object has no len()"
-> 1286                             .format(self.__class__.__name__))
   1287         return len(self.jd1)
   1288 

TypeError: Scalar Time object has no len()
```

mdmueller commented:
Yes, `Time` doesn't currently work with `loc` access but I'll find a fix--looks like the issue is with `np.searchsorted`. I'm still having some trouble determining why table slicing is about 50% slower than in 1.0.4, but most of the difference is just between master and v1.0.x rather than table-indexing and master:
```
In [1]: from astropy.table.table_helpers import simple_table

In [2]: t=simple_table(1000000)

In [3]: timeit t[1:2]
1000 loops, best of 3: 233 µs per loop
```
On master I get 210 µs, and on v1.0.x I get 182 µs. cProfile hasn't been very helpful, but I'll try a line profiler to get specifics. By the way, the huge slowdown of `t[[1,2]]` when an index is present is due to the fact that all index items are relabeled (currently in a Python loop) during list access. The relabeling is necessary, but I'm pretty sure it can be made faster.

mdmueller commented:
It seems that `searchsorted` is problematic because it turns the input array (here, a `Time` object) into an `ndarray`. I can get around this by checking for `Time` objects and passing the `.jd` attribute to `searchsorted`, though it seems like a hack and I don't know much about underlying precision issues in `Time` (it seems that I should use both `jd1` and `jd2`). I suppose that's the best way, but let me know if that's problematic with `Time`. I also wondered about storing only the `jd1` and `jd2` components of each `Time` object in the indexing engine rather than the `Time` object itself, which might make things work nicely, although then there's the issue that column values should be directly accessible via `.loc`, and we would need to designate these columns as Time columns to avoid accidental access via a tuple, or something.

taldcroft commented:
@mdmueller - I think that a generalizable way to support `Time` is by looking for a `searchsorted` method and using that if available.   You could implement a bare bones method for Time without too much trouble since it supports comparisons.

mdmueller commented:
That works, although it'll be somewhat slower than `np.searchsorted`. I'll see if I can fix the performance issue you noted above.

mhvk commented:
#3681 added support for `argsort` to `Time` -- it should be relatively easy to use it to implement `searchsorted`. But a slow placeholder should be fine for now.

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+        return False
+
+    def __repr__(self):
+        return repr(self.val) + " + epsilon"
```
Tempted to put something in here along the lines of:

```python
from astropy import conf
if conf.use_unicode:
    return repr(self.val) + " + ε"
```

:)

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+        return str(self)
+
+
+class BST(object):
```
Some of this duplicates the work in the [`ExpressionTree`](https://github.com/astropy/astropy/blob/74b1a86ac915c71193890e6452097226ee62628c/astropy/modeling/utils.py#L22) class in `astropy.modeling`, which is also a binary tree.  Of course, they serve different purposes--this one is for searching, while the `ExpressionTree` is for evaluating expressions of binary operators.  Your `BST` class is also rather different in how it uses a separate `Node` class (whereas `ExpressionTree` just nests `ExpressionTree` instances).  I think I like your approach better overall though, and might try to adapt it.

One possible advantage of the `ExpressionTree` implementation is that its traversal methods are iterative, and are implemented as generators--they don't require making a full list of nodes.  I don't *know* if that would be an advantage for this case (I haven't investigated how it's used yet), but just putting it out there.

Maybe eventually it might be worth putting a generic "BinaryTree" class somewhere, from which both `BST` and `ExpressionTree` can inherit for their own unique purposes.  It could have both iterative and recursive traversal options.  Don't bother about it right now, but I think this is worth making a note of.

mdmueller replied:
I like that last idea, especially since `bintrees` can come in handy--I can work on it once this PR is merged.

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+
+except ImportError:
+    FastBST = BST
+    FastRBT = BST
```
Ah, perhaps all the more reason to put a generic binary tree module somewhere in Astropy, that we can both use (so that we can also take advantage of the optional bintrees dependency wherever applicable).

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+    val : object
+        Original value
+    '''
+
```
I don't know how common these objects are, but it might be worth having a `__slots__ = ('val',)` for this class.

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+    __ge__ = lambda x, y: x.key >= y.key
+    __gt__ = lambda x, y: x.key > y.key
+    __ne__ = lambda x, y: x.key != y.key
+
```
This would probably also be worth using `__slots__`.

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+        self.key = other.key
+        self.data = other.data[:]
+
+    def set_left(self, node):
```
I'd just make `self.left` and self.right` properties.

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+        self.data = data if isinstance(data, list) else [data]
+        self.left = None
+        self.right = None
+        self.parent = None
```
On second thought, is the `self.parent` reference necessary?  I don't see how it's being used anywhere.  I also would worry about how the resulting reference cycles will affect garbage collection performance.  Better to avoid if it's not needed.

mdmueller replied:
It's used in `_substitute` when substituting nodes, but it looks like the substitution code can be rewritten not to use it. You're right, the impact of `parent` is probably not great (though I haven't done much memory profiling, etc.)

embray replied:
Ah okay, I noticed that it was used in `_substitute`, but I misread *how* it was being used.  This looks fine though.

embray commented on astropy/table/bst.py:
```diff
@@ -0,0 +1,614 @@
+            line += self._print(node.right, level + 1)
+        return line
+
+    def height(self):
```
This might as well be a property.

embray commented on astropy/table/index.py:
```diff
@@ -0,0 +1,781 @@
+                return index
+    return None
+
+class index_mode(object):
```
I think it would be nice actually if this were a `Table` method, so one could use it like:

```python
with my_table.index_mode('freeze'):
    # Do stuff...
```
etc.  It makes sense to me as a method on the table since it's temporarily doing something to the table's internal state.  Making it a method wouldn't require much change.  Just leave this class as is (though maybe give it a private name like `_IndexModeContext`) and add to `Table`:

```python
def index_mode(self, mode):
   return _IndexModeContext(self, mode)
```

Thoughts?

mdmueller replied:
Ah, thanks for pointing this out--we actually decided on this previously, but I forgot to implement it. I'll add a commit now.

embray replied:
Cool, thanks!

taldcroft commented:
There is something hanging up performance pretty badly in this case:
```
In [1]: from astropy.table.table_helpers import simple_table
In [2]: import numpy as np

In [3]: size = 10000
In [4]: idx = np.arange(size)
In [5]: np.random.shuffle(idx)
In [6]: t = simple_table(size)[idx]

In [7]: %time t.sort('a')
CPU times: user 5.59 ms, sys: 386 µs, total: 5.97 ms
Wall time: 5.61 ms

In [8]: t = simple_table(size)[idx]

In [9]: %time t.add_index('a')
CPU times: user 2.26 ms, sys: 498 µs, total: 2.76 ms
Wall time: 2.26 ms

In [10]: %time t.sort('a')
CPU times: user 10.4 s, sys: 17 ms, total: 10.4 s
Wall time: 10.4 s   <<<<==== YIKES!
```

mdmueller commented:
Yikes! It looks like that's the result of index relabeling due to column slicing in `sort`; there should be a quick fix since we know the indices just need to be sorted. This is related to the issue you noted with retrieving by list values, which I don't think has a nice solution. My latest commit does improve the situation dramatically for the case you tried:
```
In [1]: t = simple_table(1000000)

In [2]: t.add_index('a')

In [3]: timeit t[1:3]
1000 loops, best of 3: 252 µs per loop

In [4]: timeit t[[1, 3]]
100 loops, best of 3: 3.07 ms per loop
```
But the only improvement has been to make the replacement loop break once all rows have been replaced, so retrieval is still extremely slow when there are many list elements:
```
In [5]: idx = np.arange(1000000)

In [6]: np.random.shuffle(idx)

In [7]: timeit t[idx]
1 loops, best of 3: 1.25 s per loop
```
Unless someone has a good idea for replacing row values in an index, it seems that maybe we should just document the fact that this is a necessarily slow operation (since *row* values in an index are not in order, and we need to relabel them). If index copying is unnecessary, there's always the 'discard_on_copy' mode.

embray commented on astropy/table/column.py:
```diff
@@ -301,6 +308,32 @@ def __reduce__(self):
 
         return reconstruct_func, reconstruct_func_args, state
 
+    def get_item(self, item):
```
Depending on whether or not people like #4075, if we merge it it might obviate the need for these methods.  With #4075 I was able to almost completely eliminate overhead related to overriding `__getitem__` in `Column` (at least for the basic case--`MaskedColumn.__getitem__` is still slowish due to the complexities of masked arrays, though in my PR it still adds less overhead than it did before).

And of course there's some overhead related to performing additional checks in the `__getitem__` implementation, but with some cleverness that can almost be completely eliminated too.

embray commented on astropy/table/column.py:
```diff
@@ -301,6 +308,32 @@ def __reduce__(self):
+            return self.data[item]  # Return as plain ndarray or ma.MaskedArray
+
+        col_slice = super(BaseColumn, self).__getitem__(item)
+        if isinstance(col_slice, BaseColumn):
```
I think in this case we can get away with writing just `if col_slice.__class__ is self.__class__`.  Normally I'd turn my nose up at at that.  But the purpose of this check if I understand correctly is that a slice or some other type of mask was taken of a Numpy array (or subclass thereof).  So I think we can make the assumption that that will always return an instance of the same class that the slice/mask was taken from (unless we also need to support some kind of mixin column where that assumption doesn't hold?  But in that case the mixin could override this code).

In any case, assuming we can get away with it, using `is` will be much faster here.

embray commented on astropy/table/index.py:
```diff
@@ -0,0 +1,781 @@
+                if isinstance(col, Time):
+                    new_columns.append(col.jd)
+                    remainder = col - col.__class__(col.jd, format='jd')
+                    new_columns.append(remainder.jd)
```
I don't think worth bothering with right now, but it might be nice if an interface were added to mixin columns that allow them to provide a view or transformation of the values in the column for use in lexical sorting.

embray commented on astropy/table/index.py:
```diff
@@ -0,0 +1,781 @@
+    def __init__(self, columns, engine=None, data=None):
+        from .table import Table
+
+        if data is not None: # create from data
```
Since in this case the engine is pre-determined anyways, it seems to me we could do away with the separate `data` argument (I can't find any code that uses it yet?), and allow `engine` to accept a type or an instance.  If `not isinstance(engine, type)` then we would do the same as in this clause, and otherwise instantiate the engine with the row data like below.

On the other hand, I like the symmetry of this with the attributes of the same names.  It just seems a little confusing to me (especially considering that you can also pass in something completely different as `engine`, but it's ignored if `data is not None`).

mdmueller replied:
You're right, using just `engine` makes more sense. I think `data` is a relic from some earlier code.

embray commented on astropy/table/index.py:
```diff
@@ -0,0 +1,781 @@
+                    self.columns.append(col)
+                elif isinstance(col, MaskedColumn):
+                    col.__class__ = _GetitemMaskedColumn
+                    self.masked_columns.append(col)
```
Not that this isn't fine for now, but I also think #4075 removes the need for this temporary class switcheroo, and instead can just tweak the appropriate flags to control the behavior.

embray commented on astropy/table/column.py:
```diff
@@ -1091,28 +1131,48 @@ def insert(self, obj, values, mask=None):
-    def __getitem__(self, item):
-        out = super(MaskedColumn, self).__getitem__(item)
-
+    def copy_attribs(self, out):
```
This doesn't seem like it should be a public method.  Maybe also call it `_copy_attrs_slice` or something to emphasize its primary context.

embray commented:
@taldcroft @mdmueller Out of curiosity, I tried rebasing this branch on top of my work in #4075.  This allowed me to get rid of the `get_item` methods altogether, as the Cython/C implementation of `Column.__getitem__` is almost imperceptibly less fast than the default implementation (when not taking a slice, that is).  You can see this work here: https://github.com/embray/astropy/tree/table-indexing  Though the commit of particular interest is 07e506056ea3fe6c663e90f3b44567342c795e1f

As I noted in the commit message, this makes `table.index_mode('copy_on_getitem')` a no-op, since that's now the default behavior.  I don't know whether or not you *want* that to be the default behavior, or if the fact that it wasn't was just for performance considerations in overriding `__getitem__` (that is, are you okay with slicing being slowed down in general?)

I could see it either way, but in any case this mostly solves the performance problems with overriding `__getitem__`.

mdmueller commented:
That sounds good to me, unless there are other reasons not to override `__getitem__`. I was under the impression that we were avoiding the override only for performance reasons.

embray commented:
@mdmueller Much of the performance impact of overriding `__getitem__` in a Python class comes merely from the fact that it requires a dict lookup every time you index the column (or any other object that overrides `__getitem__`).  That is, a lot of the slowdown comes from looking up the `__getitem__` implementation itself (independent of how that implementation works, which may or may not be slow).

Because Numpy arrays are built-in classes (that is, they're implemented in C), running `array[idx]` involves nothing more than a few pointer dereferences: `Py_TYPE(obj)->tp_as_mapping->mp_subscript`

Subclasses of `ndarray` will automatically get the `ndarray` class's slots copied, so by default `__getitem__` on subclasses is just as fast.  However, if a subclass *overrides` `__getitem__`, then you get the laborious process of lookup in the class dict, along with all the trappings of normal attribute access like descriptor support.  This alone results in a roughly 8x slowdown even for a trivial `__getitem__` *implementation*.

Overriding `__getitem__` *could* still end up being a problem even if it's written in C like in #4075, but it depends on what we're doing in the implementation.  If we can pass through to the basic implementation for the simple case of a single scalar lookup with only a few checks then the performance impact is negligible.

For something like taking a slice, however, the impact is not completely negligible if we decide to have it copy the indices by default.  I don't know if that matters or not though--I don't think we typically care as much about the performance impact on slice.

taldcroft commented:
@embray - this new way of improving the column `__getitem__` does look nicer and I'm all for going in that direction.

For now we probably want just work on getting the table indexing PR merged and then deal with cleanup later.

embray commented:
@taldcroft That's fine--let's get this finished and merged first.  I can later take my branch and spin it into a PR.

mdmueller commented:
I just noticed something that seems to be a slight issue for `.loc`; currently a call like `t.loc[val]` returns an empty table if `val` is not found in the indexed column. But for `Time` objects, this might not be what we want, since `Time` cannot be empty:
```
In [4]: t = Table([Time([100, 200, 300], format='jyear')])

In [5]: empty_table = t[[]]

In [6]: empty_table
Out[6]: 
<Table length=0>
 col0 
object
------

In [7]: empty_table['col0']
Out[7]: ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-d8887fc8ab04> in <module>()
----> 1 empty_table['col0']

/usr/lib/python2.7/dist-packages/IPython/core/displayhook.pyc in __call__(self, result)
    245             self.start_displayhook()
    246             self.write_output_prompt()
--> 247             format_dict, md_dict = self.compute_format_data(result)
    248             self.write_format_data(format_dict, md_dict)
    249             self.update_user_ns(result)

/usr/lib/python2.7/dist-packages/IPython/core/displayhook.pyc in compute_format_data(self, result)
    155 
    156         """
--> 157         return self.shell.display_formatter.format(result)
    158 
    159     def write_format_data(self, format_dict, md_dict=None):

/usr/lib/python2.7/dist-packages/IPython/core/formatters.pyc in format(self, obj, include, exclude)
    150             md = None
    151             try:
--> 152                 data = formatter(obj)
    153             except:
    154                 # FIXME: log the exception

/usr/lib/python2.7/dist-packages/IPython/core/formatters.pyc in __call__(self, obj)
    479                 type_pprinters=self.type_printers,
    480                 deferred_pprinters=self.deferred_printers)
--> 481             printer.pretty(obj)
    482             printer.flush()
    483             return stream.getvalue()

/usr/lib/python2.7/dist-packages/IPython/lib/pretty.pyc in pretty(self, obj)
    360                             if callable(meth):
    361                                 return meth(obj, self, cycle)
--> 362             return _default_pprint(obj, self, cycle)
    363         finally:
    364             self.end_group()

/usr/lib/python2.7/dist-packages/IPython/lib/pretty.pyc in _default_pprint(obj, p, cycle)
    480     if getattr(klass, '__repr__', None) not in _baseclass_reprs:
    481         # A user-provided repr.
--> 482         p.text(repr(obj))
    483         return
    484     p.begin_group(1, '<')

/home/michael/git/astropy/astropy/time/core.py in __repr__(self)
    390         return ("<{0} object: scale='{1}' format='{2}' value={3}>"
    391                 .format(self.__class__.__name__, self.scale, self.format,
--> 392                         getattr(self, self.format)))
    393 
    394     def __str__(self):

/home/michael/git/astropy/astropy/time/core.py in __getattr__(self, attr)
   1077             if hasattr(self.FORMATS[attr], 'epoch_scale'):
   1078                 tm._set_scale(self.FORMATS[attr].epoch_scale)
-> 1079             return tm.value
   1080 
   1081         elif attr in TIME_SCALES:  # allowed ones done above (self.SCALES)

/home/michael/git/astropy/astropy/time/core.py in value(self)
    555     def value(self):
    556         """Time value(s) in current format"""
--> 557         return self._shaped_like_input(self._time.value)
    558 
    559     def sidereal_time(self, kind, longitude=None, model=None):

/home/michael/git/astropy/astropy/time/core.py in value(self)
   2551     def value(self):
   2552         jd_to_epoch = getattr(erfa_time, self.jd_to_epoch)
-> 2553         return jd_to_epoch(self.jd1, self.jd2)
   2554 
   2555 

/home/michael/git/astropy/astropy/_erfa/core.pyc in epj(dj1, dj2)
  22161     op_axes = [[-1]*(broadcast.nd-arr.ndim) + list(range(arr.ndim)) for arr in arrs]
  22162     op_flags = [['readonly']]*2 + [['readwrite']]*1
> 22163     it = numpy.nditer(arrs, op_axes=op_axes, op_flags=op_flags)
  22164 
  22165     #Iterate

ValueError: Iteration of zero-sized operands is not enabled
```
The error occurs even during table slicing when I modify the index slicing behavior to simply recreate an index (which is markedly faster in certain cases). My question is: should we guard against returning an empty table in `.loc` (e.g. by throwing an exception if no value is found), or should we look for a workaround so that we can continue to return some sort of empty table?

taldcroft commented:
I think that `.loc` should raise an exception if the `item` is not found.  This method is basically akin to key access, so the exception should probably be `KeyError`.  That follows the pandas precedent as well.  For `.iloc` accessing a non-existent item would be an `IndexError` as expected.

taldcroft commented:
In the list of features at the top of the Table index page, add a line (after the database operations one):
```
- Maintain a table index for fast retrieval of table items or ranges.
```

taldcroft commented:
Unrelated, but while you are there why don't you change ".. notable features of this package .." to ".. notable capabilities of this package ..".  I think the latter word fits slightly better.

taldcroft commented on docs/table/indexing.rst:
```diff
@@ -0,0 +1,219 @@
+To create an index on a table, use the |add_index| method::
+
+   >>> from astropy.table import Table
+   >>> t = Table([(1, 2, 3, 4), (5, 6, 7, 8)], names=('a', 'b'))
```
This first example should highlight how things get sorted, and the possibility of non-unique keys for column 'a'.  So use something like:
```
>>> t = Table([(2, 3, 2, 1), (8, 7, 6, 5)], names=('a', 'b'))```

taldcroft commented on docs/table/indexing.rst:
```diff
@@ -0,0 +1,219 @@
+
+   >>> t = Table([(1, 2, 3, 4), (10, 1, 9, 9)], names=('a', 'b'))
+   >>> t.add_index('a')
+   >>> t.loc[2]
```
After our discussion at the final hangout, I think this should be returning a `Row` whenever the result has length=1.

mdmueller commented:
Sounds good--I made these changes in the latest commit.

One performance issue I still can't figure out is the behavior of `FastBST` for `loc` range queries, even when index copying is disabled. cProfile tracing indicates that the time is pretty much spent in `bintrees` code, but it seems really strange to me that even in O(n) worst-case time, an optimized C-based tree is slower than an ordinary Python search. Either there's something subtle going on or bintrees has a real issue...I'm just taking note of this for investigation after this PR is merged.

taldcroft commented:
@mdmeuller @mdboom - have either of you had a chance to run the current asv table tests on this branch to see if there are any unexpected / serious performance regressions?

taldcroft commented:
Indexing.rst failing because the row repr is not consistent with the new behavior post-#4066.

taldcroft commented:
Hi @mdmueller, I'm guessing you got busy with the new school year.  For the 1.1 release the feature freeze is on Monday which means getting this merged or really close by then.  What needs finishing up is:

- [ ] Rebase
- [ ] Fix the test failures (which I think are minor)
- [ ] Run this branch against the time_table benchmarks in `astropy-benchmarks`, make sure there are no major performance regressions vs. master
- [ ] Add a warning note in the indexing docs that this release of indexing is new and not considered stable yet.
- [ ] Move the ipython notebook with your indexing performance into `astropy-notebooks`

If you can't get to these in the next few days we may have to push off to 1.2.  This wouldn't be the end of the world because we could get it into master early in the release cycle and get a lot more exposure before release.  Either way...

mdmueller commented:
Oops, sorry about the delay! I've been busy as you guessed, but I did mean to get to this sooner.

I've rebased and added a warning note. I'm having trouble with asv on my current system but here's the result of an informal profile:
```
** master **

time_add_column                    0.00576972961426
time_add_row                       0.0191135406494
WARNING: Cannot create a consistent method resolution
order (MRO) for bases Column, _NDColumnProxyShim [astropy.table.groups]
time_aggregate                     0.0397601127625
time_aggregate_noreduceat          0.620079040527
time_column_get                    5.3882598877e-05
time_column_make_bool_mask         0.00154972076416
time_column_set                    0.000261068344116
time_column_set_all                0.000125885009766
time_column_set_row_subset         0.000245332717896
time_column_set_row_subset_int     0.000183820724487
time_column_slice_bool             0.00150227546692
time_column_slice_int              0.00188994407654
time_copy_column                   0.00195407867432
time_copy_table                    0.0171203613281
time_group                         0.0683679580688
time_hstack                        0.0531780719757
time_init_from_np_array_copy       0.0108757019043
time_init_from_np_array_no_copy    0.00598788261414
time_item_get_colfirst             5.81741333008e-05
time_item_get_rowfirst             0.000288009643555
time_iter_row                      0.16452050209
time_join_inner                    0.0876154899597
time_join_outer                    0.195106267929
time_multi_column_get              0.00946879386902
time_read_rows                     0.341752767563
time_remove_column                 0.000110864639282
time_remove_row                    0.0109417438507
time_remove_rows                   0.0105555057526
time_row_get                       0.00113201141357
time_sort                          0.0356187820435
time_table_slice_bool              0.011932849884
time_table_slice_int               0.0120806694031
time_vstack                        0.0367367267609

** table-indexing **

time_add_column                    0.0053186416626 (92.2%)
time_add_row                       0.0213468074799 (111.7%)
WARNING: Cannot create a consistent method resolution
order (MRO) for bases Column, _NDColumnProxyShim [astropy.table.groups]
time_aggregate                     0.0419991016388 (105.6%)
time_aggregate_noreduceat          0.62233710289 (100.4%)
time_column_get                    5.91278076172e-05 (109.7%)
time_column_make_bool_mask         0.00210785865784 (136.0%)
time_column_set                    0.000347137451172 (133.0%)
time_column_set_all                0.000251531600952 (199.8%)
time_column_set_row_subset         0.000346183776855 (141.1%)
time_column_set_row_subset_int     0.000270128250122 (146.9%)
time_column_slice_bool             0.00165104866028 (109.9%)
time_column_slice_int              0.00154280662537 (81.6%)
time_copy_column                   0.00205230712891 (105.0%)
time_copy_table                    0.016015291214 (93.5%)
time_group                         0.0754358768463 (110.3%)
time_hstack                        0.0601060390472 (113.0%)
time_init_from_np_array_copy       0.0129377841949 (119.0%)
time_init_from_np_array_no_copy    0.00610065460205 (101.9%)
time_item_get_colfirst             6.53266906738e-05 (112.3%)
time_item_get_rowfirst             0.000327825546265 (113.8%)
time_iter_row                      0.151817083359 (92.3%)
time_join_inner                    0.0958755016327 (109.4%)
time_join_outer                    0.199917316437 (102.5%)
time_multi_column_get              0.00955867767334 (100.9%)
time_read_rows                     0.333776950836 (97.7%)
time_remove_column                 0.000125646591187 (113.3%)
time_remove_row                    0.0127828121185 (116.8%)
time_remove_rows                   0.0133459568024 (126.4%)
time_row_get                       0.000250339508057 (22.1%)
time_sort                          0.0433120727539 (121.6%)
time_table_slice_bool              0.0137088298798 (114.9%)
time_table_slice_int               0.0118918418884 (98.4%)
time_vstack                        0.0395669937134 (107.7%)
```

Mainly there's a slowdown in column setting (which seems a little strange to me, I can look into this at some point).

astrofrog commented:
@taldcroft @embray @mdmueller @mdboom - what is the status of this PR? Can it be merged today for 1.1 or are the performance issues blockers?

embray commented:
Note: If this is merged now, I can easily also incorporate my changes from #4075 which will simplify some aspects of this PR (no `get_item` method) and address some of the performance issues.

I don't think #4075 will help with the column set slowdowns, which are unsurprising.  I think the column set slowdowns are probably acceptable since it's a less common operation and less performance-critical in practice.  @taldcroft has the final call on that though.

IMO all the checks are passing on this PR; I think we should merge it.

embray commented:
(At least after including a changelog entry ;)

taldcroft commented:
I'm go for merge.  I don't think the performance hits are too bad in the common cases.  @embray - if you want to merge and get going on #4075 right away (and deal with the changelog separately) then that's fine.

astrofrog commented:
@mdmueller - it looks like there is a doctest that needs skipping on Windows (see the AppVeyor 2.6 failure)

taldcroft commented:
@mdmueller - instead of skipping the doctest you can add the type (and check that all your examples specify int64 where needed).
```
t = Table([(1, 2, 3, 4), (10, 1, 9, 9)], names=('a', 'b'), dtype=np.int64)
```

taldcroft commented:
To try finishing this out and to provide a passing base for #4090, I've cloned this PR off to #4202 and added a commit to fix the Windows docstring problem.  If that passes then I'll merge #4202.

taldcroft commented:
Closed by #4202.
